### DISCUSSION TITLE
Multiple errors for one field

### DISCUSSION THREAD

**Discussion by trickreich** [react-hook-form/react-hook-form#7060] [Q&A] [answered]
I'm not sure how to handle multiple errors for one field.
Actually I'm using `setError` and pass an `Array<string>` instead of a `string` for `message`.
Sound's this legit to someone else?

```
form.setError("myfield", {
    message: ["Error Message 1", "Error Message",
})
```

**Comment by bluebill1049** [comment]
https://react-hook-form.com/api/useform

`criteriaMode` ?

**Reply by trickreich** [reply]
Hmm.. not sure if this could fit in my case.
I'm getting back some errors from my request and need to set all errors to the form object.

**Reply by bluebill1049** [reply]
oh, i see, an error message is only returned as a string. if you need to return an array of messages perhaps you want to use React.useState instead.

**Reply by bluebill1049** [reply]
you can return types tho, each type can be associated with one message

```tsx
setError('test', {
  types: {
    data: 'test',
    data1: 'test'
  }
})
```

**Reply by trickreich** [reply]
my hack with an array in `message` is working.
any disadvantages for us this way?

**Reply by trickreich** [reply]
is this api with `types` a misstype or not documented?
https://react-hook-form.com/api/useform/seterror

**Reply by bluebill1049** [reply]
oh... oops will get that updated

**Reply by bluebill1049** [reply]
an example is there.
<img width="1047" alt="Screen Shot 2021-11-17 at 9 09 35 am" src="https://user-images.githubusercontent.com/10513364/142074023-80cd338a-672f-4b6e-83cd-f00811acd50d.png">

**Reply by trickreich** [reply]
also it wasn't easy for me to get which type I could use, didn't know it's just a free string

**Comment by trickreich** [comment, marked as answer]
I've done following to handle multiple errors from a api response:

```
function setMultipleErrors(field, errors) {
    const types = {}
    errors.forEach((error, i) => {
        types[`apiError${i + 1}`] = error
    })

    form.setError(field, {
        types,
    })
}

function handleApiErrors(apiErrors) {
    const newFormErrors = {}

    apiErrors.forEach(({ field, messages }) => {
        messages.forEach((message) => {
            if (field in form.getValues()) {
                if (!newFormErrors[field]) {
                    newFormErrors[field] = []
                }

                newFormErrors[field].push(message)
            } else {
                if (!newFormErrors.__general__) {
                    newFormErrors.__general__ = []
                }

                newFormErrors.__general__.push(message)
            }
        })
    })

    Object.entries(newFormErrors).forEach(([field, errors]) =>
        setMultipleErrors(field, errors),
    )
}
```

I've introduced a virtual field named `__general__` to save errors, when the form object doesn't know the field.

What you guys think about such an approach?

**Reply by bluebill1049** [reply]
I think that's fine.

**Reply by Moshyfawn** [reply]
On the topic of introducing an additional field, you can take a look at #7661 and vote for a global errors fields feature

**Reply by vairadurai007** [reply]
give me the solution to the how to set multiple error to the multiple email validation
